WalletWasabi.Gui/Tabs/WalletManager/LoadWalletViewModel.cs

```diff
@@ -440,122 +394,109 @@ public async Task<KeyManager> LoadKeyManagerAsync(bool requirePassword, bool isH
 				var walletName = selectedWallet.WalletName;
 				if (isHardwareWallet)
 				{
-					if (selectedWallet.HardwareWalletInfo is null)
+					using (var cts = new CancellationTokenSource(TimeSpan.FromMinutes(3)))
 					{
-						SetValidationMessage("No hardware wallets detected.");
-						return null;
-					}
+						var client = new HwiClient(Global.Network);
 
-					if (!selectedWallet.HardwareWalletInfo.Initialized)
-					{
-						IEnumerable<HardwareWalletInfo> hwis;
-						try
+						if (selectedWallet.HardwareWalletInfo is null)
 						{
-							IsHardwareBusy = true;
-							MainWindowViewModel.Instance.StatusBar.TryAddStatus(StatusBarStatus.SettingUpHardwareWallet);
-							if (!await HwiProcessManager.SetupAsync(selectedWallet.HardwareWalletInfo))
+							SetValidationMessage("No hardware wallets detected.");
+							return null;
+						}
+
+						if (!selectedWallet.HardwareWalletInfo.IsInitialized())
+						{
+							try
+							{
+								IsHardwareBusy = true;
+								MainWindowViewModel.Instance.StatusBar.TryAddStatus(StatusBarStatus.SettingUpHardwareWallet);
+
+								// Setup may take a while for users to write down stuff.
+								using (var ctsSetup = new CancellationTokenSource(TimeSpan.FromMinutes(21)))
+								{
+									// Trezor T doesn't require interactive mode.
+									if (selectedWallet.HardwareWalletInfo.Model == HardwareWalletModels.Trezor_T
+									|| selectedWallet.HardwareWalletInfo.Model == HardwareWalletModels.Trezor_T_Simulator)
+									{
+										await client.SetupAsync(selectedWallet.HardwareWalletInfo.Model, selectedWallet.HardwareWalletInfo.Path, false, ctsSetup.Token);
+									}
+									else
+									{
+										await client.SetupAsync(selectedWallet.HardwareWalletInfo.Model, selectedWallet.HardwareWalletInfo.Path, true, ctsSetup.Token);
+									}
+								}
+
+								MainWindowViewModel.Instance.StatusBar.TryAddStatus(StatusBarStatus.ConnectingToHardwareWallet);
+								await EnumerateHardwareWalletsAsync();
```

Comment:
I don't remember if I already says this but imo this method should accept a cancellation token.

Reply:
Same reasoning as in the similar question. We seemed to found 3m as the "sweetspot" for timeout. Exposing it would risk to introduce future bugs. https://github.com/zkSNACKs/WalletWasabi/issues/2312